### DISCUSSION TITLE
Embedding CVAT into a Webservice

### DISCUSSION THREAD

**Discussion by johanneszellinger** [cvat-ai/cvat#4526] [Q&A]
Hello, 

we are interested in embedding CVAT into a custom web service.
The idea is that we use the REST API to generate all the task/projects, upload data or run models to pre-annotate data. 
The user should only have the labelling GUI to work with, which could be shown simply within an <iframe> tag in our HTML. 

Some questions came up:
- There are some errors when accessing the data, which - just a wild guess - could be related to CORS. Is there a way to enable CORS? We are running CVAT in a docker-container (set up as described [in the docs](https://openvinotoolkit.github.io/cvat/docs/administration/basics/installation/#windows-10)). 
- Is there a way to hide parts of the UI in CVAT. For example: we want to control the user-login via the REST API, the user shouldn't even have the option to log in or out. 
- Additionally, depending on the task, we want to hide some specific options. E.g for a detection task we might want the user only to draw quadratic bounding boxes and remove the option for ellipsis or polygons. 

Any help/suggestion is much appreciated!

**Comment by nmanovic** [comment]
@TheRealRolandDeschain , sure. We will be happy to collaborate and help. I know that many teams did the exercise but unfortunately nobody decided to contribute and document the integration. If you can help in the area, community will be happy and we will try to support the solution in future releases.

* You need to modify settings: https://github.com/openvinotoolkit/cvat/blob/6fd7a0e1aef17793d41a2362c0f8db53d9d83eac/cvat/settings/development.py#L43 and also look at documentation for Django CORS headers package: https://pypi.org/project/django-cors-headers/
* We don't have such possibility at the moment. But we are happy to accept a PR with such changes.
* It is a more complex feature. Basically there are several ways to solve the issue:
  * extend definition of a task and project on the server with extra fields (e.g., task_type: str) and use the information in UI to hide some parts of UI.
  * extend the labels scheme and add information about expected shapes for the label. Thus we can use the information to hide some UI as well (e.g., if only bounding boxes are necessary for all labels). The approach is more flexible and always can cover needs for the first approach.

**Reply by johanneszellinger** [reply]
Thanks I'll have a look into it the coming days!

**Reply by johanneszellinger** [reply]
@nmanovic sorry for bothering you again, but a question came up. 

I think the issue in our case was not CORS, but Same Origin Policy. I guess the solution for this would be to set up CVAT behind a reverse proxy. I have already implemented such a reverse proxy on our webserver in the following way:

- http://localhost:8080 is the CVAT address
- https://localhost:7123 is the address of our server 
- the reverse-proxy currently maps https://localhost:7123/cvat (or rather all requests on our webserver which have the pattern `/cvat/{**catchall}` in them) to http://localhost:8080. 

The issue I'm facing is, that all static files, such as css, js,.. which are linked in the return html are not properly mapped, resulting for example in request to `https://localhost:7123/assets/cvat-ui.8aefd0b1545e6a819b47.min.js` which will simply result in a 404 error - obviously. 
In the CVAT documentation there is a section on[ how to set up CVAT behind a proxy](https://openvinotoolkit.github.io/cvat/docs/administration/basics/installation/#deploying-cvat-behind-a-proxy), but I'm not quite sure how to correctly set this up in my case. 
For example, adding an `.env` file with the following environment variables would not work:

> HTTPS_PROXY=https://localhost:7123
> HTTP_PROXY=https://localhost:7123

I'm happy to add a detailed description of the whole exercise to the documentation as soon as we manage to get it up and running, however at the moment I'm still failing myself 😄

**Reply by nilsmelchert** [reply]
@johanneszellinger Did you succeed with the exercise? 😉 
I would also be very interested in that solution.

**Reply by johanneszellinger** [reply]
@nilsmelchert Unfortunately not really. In my project we fixed at least the routing issues by putting CVAT on a kubernetes cluster behind an Nginx server. A colleague of mine did a project with CVAT which was  a annotation training station designed for people with autism (unfortunately only available in German: https://www.trainingstation.at/) where people could test their abilities to annotate and maybe get a fairly paid job. As far as I know it is a vuejs app, where vue logic handles authorization tokens and the CVAT UI is embedded with parts of the UI hidden as well by bue logic - so in short this works, but is a bit hacky^^
